File: RoydeWildt/QLJava/src/org/uva/sea/ql/ast/visitor/Visitor.java
package org.uva.sea.ql.ast.visitor;

import org.uva.sea.ql.ast.tree.form.*;
import org.uva.sea.ql.ast.tree.stat.*;
import org.uva.sea.ql.ast.tree.expr.unary.*;
import org.uva.sea.ql.ast.tree.expr.binary.*;
import org.uva.sea.ql.ast.tree.val.*;


/**
 * Created by roy on 5-2-16.
 */
public interface Visitor {
    Visitor v = null;
    <T> T visit(Form form);

    <T> T visit(If stat);
    <T> T visit(IfElse stat);
    <T> T visit(Question stat);
    <T> T visit(AssQuestion stat);

    <T> T visit(Add expr);
    <T> T visit(And expr);
    <T> T visit(Div expr);
    <T> T visit(Eq expr);
    <T> T visit(GEq expr);
    <T> T visit(GT expr);
    <T> T visit(LEq expr);
    <T> T visit(LT expr);
    <T> T visit(Mul expr);
    <T> T visit(NEq expr);
    <T> T visit(Or expr);
    <T> T visit(Sub expr);
    <T> T visit(Neg expr);
    <T> T visit(Not expr);
    <T> T visit(Pos expr);
    <T> T visit(Primary expr);

    <T> T visit(Bool  val);
    <T> T visit(Int  val);

    <T> T visit(Var var);

}

File: Edward Poot/src/ql/interfaces/visitor.go
package interfaces

type Visitor interface {
	/* Expr */
	VisitAdd(Add, interface{}) interface{}
	VisitAnd(And, interface{}) interface{}
	VisitDiv(Div, interface{}) interface{}
	VisitEq(Eq, interface{}) interface{}
	VisitGEq(GEq, interface{}) interface{}
	VisitGT(GT, interface{}) interface{}
	VisitLEq(LEq, interface{}) interface{}
	VisitLT(LT, interface{}) interface{}
	VisitMul(Mul, interface{}) interface{}
	VisitNEq(NEq, interface{}) interface{}
	VisitOr(Or, interface{}) interface{}
	VisitSub(Sub, interface{}) interface{}
	VisitBoolLiteral(BoolLiteral, interface{}) interface{}
	VisitIntegerLiteral(IntegerLiteral, interface{}) interface{}
	VisitStringLiteral(StringLiteral, interface{}) interface{}
	VisitNeg(Neg, interface{}) interface{}
	VisitNot(Not, interface{}) interface{}
	VisitPos(Pos, interface{}) interface{}
	VisitVarExpr(VarExpr, interface{}) interface{}
	/* Vari */
	VisitVarDecl(VarDecl, interface{}) interface{}
	VisitVarID(VarID, interface{}) interface{}
	VisitIntegerType(IntegerType, interface{}) interface{}
	VisitBoolType(BoolType, interface{}) interface{}
	VisitStringType(StringType, interface{}) interface{}
	/* Stmt */
	VisitForm(Form, interface{}) interface{}
	VisitComputedQuestion(ComputedQuestion, interface{}) interface{}
	VisitInputQuestion(InputQuestion, interface{}) interface{}
	VisitIf(If, interface{}) interface{}
	VisitIfElse(IfElse, interface{}) interface{}
	VisitStmtList(StmtList, interface{}) interface{}
}
